### PR TITLE
Allow embedders to schedule a callback on all engine managed threads.

### PR DESCRIPTION
**Pull request by chinmaygarde**
`FlutterEnginePostCallbackOnAllNativeThreads` schedule a callback to be run on
all engine managed threads. The engine will attempt to service this callback the
next time the message loops for each managed thread is idle. Since the engine
manages the entire lifecycle of multiple threads, there is no opportunity for
the embedders to finely tune the priorities of threads directly, or, perform
other thread specific configuration (for example, setting thread names for
tracing). This callback gives embedders a chance to affect such tuning.

Fixes https://github.com/flutter/flutter/issues/49551
Fixes b/143774406
Fixes b/148278215
Fixes b/148278931

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @dnicoara

File: shell/platform/embedder/embedder.h [hunk 1]
```diff
@@ -934,6 +934,31 @@ typedef struct {
   };
 } FlutterEngineDartObject;
 
+/// This enum allows the embedders to determine the type of the engine thread in
```
Comment (cbracken) [inline shell/platform/embedder/embedder.h]: uber-pedantic-nit: Can save a word by doing `s/allows the embedders/allows embedders/` -- it's also entirely fine as-is though :)

Comment (chinmaygarde) [reply shell/platform/embedder/embedder.h]: Done.

File: shell/platform/embedder/embedder.h [hunk 2]
```diff
@@ -1667,6 +1692,45 @@ FLUTTER_EXPORT
 FlutterEngineResult FlutterEngineNotifyLowMemoryWarning(
     FLUTTER_API_SYMBOL(FlutterEngine) engine);
 
+//------------------------------------------------------------------------------
+/// @brief      Schedule a callback to be run on all engine managed threads.
+///             The engine will attempt to service this callback the next time
+///             the message loops for each managed thread is idle. Since the
```
Comment (cbracken) [inline shell/platform/embedder/embedder.h]: s/loops/loop/

Comment (chinmaygarde) [reply shell/platform/embedder/embedder.h]: Done.

File: shell/platform/embedder/embedder.h [hunk 3]
```diff
@@ -1667,6 +1692,45 @@ FLUTTER_EXPORT
...
+///             so may risk performance issues (especially for callbacks of type
+///             kFlutterNativeThreadTypeUI and kFlutterNativeThreadTypeRender).
+///
+/// @attention  Some callbacks (especially the ones of type
+///             kFlutterNativeThreadTypeWorker) may be called after the
+///             FlutterEngine instance has shut down. Embedders must be careful
+///             in handling the lifecycle of objects associated with the user
+///             data baton.
+///
+/// @attention  In case there are multiple running Flutter engine instance,
```
Comment (cbracken) [inline shell/platform/embedder/embedder.h]: nit: s/instance/instances/

Comment (chinmaygarde) [reply shell/platform/embedder/embedder.h]: Done.

File: shell/platform/embedder/embedder.h [hunk 4]
```diff
@@ -1667,6 +1692,45 @@ FLUTTER_EXPORT
...
+///             their workers are shared.
+///
+/// @param[in]  engine     A running engine instance.
+/// @param[in]  callback   The callback that will get called multiple times on
+///                        each engine managed thread.
+/// @param[in]  user_data  A baton passed by the engine to the callback. This
+///                        baton is not interpreted by the engine in any way.
+///
+/// @return     The flutter export.
```
Comment (jason-simmons) [inline shell/platform/embedder/embedder.h]: Change this to "The result of the call"

File: shell/platform/embedder/embedder.cc [hunk 5]
```diff
@@ -1747,3 +1746,27 @@ FlutterEngineResult FlutterEngineNotifyLowMemoryWarning(
                    kInternalInconsistency,
                    "Could not dispatch the low memory notification message.");
 }
+
+FlutterEngineResult FlutterEnginePostCallbackOnAllNativeThreads(
```
Comment (cbracken) [inline shell/platform/embedder/embedder.cc]: Question -- I didn't look through the two bugs immediately before reviewing but what's the rationale for posting to all threads and having the user potentially do something conditionally on the `type` param in the closure rather than having this take a thread-type and posting only to the specified task runner? Is the use-case such that you want to post on all or none?

Comment (cbracken) [reply shell/platform/embedder/embedder.cc]: Ah I see -- so we can provide some guarantee of synchronicity, looks like.

Comment (chinmaygarde) [reply shell/platform/embedder/embedder.cc]: The use case requires posting to all threads. But, I can see your approach as being viable as well. Posting to specific threads repeatedly is not a use case that we want to handle. This is supposed to be called once very early on in engine lifecycle.

File: shell/platform/embedder/embedder.h (hunk 4 above)
Comment (chinmaygarde) [reply shell/platform/embedder/embedder.h]: Done.

File: shell/platform/embedder/embedder.cc (hunk 5 above)
Comment (cbracken) [reply shell/platform/embedder/embedder.cc]: Should we be warning people not to attempt to do any cross-thread latching in the closures they pass in?

**Comment by chinmaygarde** [comment]
hmm, I just realized you cannot post a callback from a callback. I’ll make that safe and test it before committing this.

**Comment by dnicoara** [comment]
Wondering, would this be sufficient to cover corner cases? Would it be possible for the embedder to call FlutterEnginePostCallbackOnAllNativeThreads before all the threads are created? With respect to worker threads, would it be possible for new worker threads to be spawned at arbitrary points in time?

**Comment by chinmaygarde** [comment]
> Would it be possible for the embedder to call FlutterEnginePostCallbackOnAllNativeThreads before all the threads are created?

No. If you have an engine ref, this is guaranteed to work.

> With respect to worker threads, would it be possible for new worker threads to be spawned at arbitrary points in time?

The number of worker threads are constant at this time. If/when this changes, the engine can hold onto the callback till the engine instance is collected. Maybe I should document that the callback may be executed a long time into the future and the embedder should not associate any handles with the baton with the expectation that all callbacks will be made in a timely manner?

**Comment by chinmaygarde** [comment]
I have patched and tested [the re-entrancy issue](https://github.com/flutter/engine/pull/15980#issuecomment-578357775).

**Comment by chinmaygarde** [comment]
The Windows presubmit failure is MSVC "fatal error C1001: An internal error has occurred in the compiler.". Will try to simplify the statement.

**Comment by dnicoara** [comment]
Awesome, thank you for quelling my concerns!
